File: .NET Core/Api/ApiClient.cs
using com.contidio.sdk.proto;
using Contidio.Sdk.Exceptions;
using System;
using System.Globalization;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading;
using System.Threading.Tasks;

namespace Contidio.Sdk.Api
{
    public class ApiClient : IDisposable         
    {
        public enum Method
        {
            GET,
            PUT,
            POST,
            DELETE
        }

        private HttpClient _httpClient;

        private bool _handleAsError;
        private HttpStatusCode _errorStatusCode;
        private string _errorReason;
        private string _bearerToken;        

        private ApiEnvironment _environment;

        private string _subscriptionKey;

        public static ApiEndpoint ENDPOINT_WWW = new ApiEndpoint(true, "www", 31442, true);        
        public static ApiEndpoint ENDPOINT_ID = new ApiEndpoint(true, "id", 31443, false);
        public static ApiEndpoint ENDPOINT_CW = new ApiEndpoint(true, "cw", 31444, false);
        public static ApiEndpoint ENDPOINT_CR = new ApiEndpoint(true, "cr", 31444, false);
        public static ApiEndpoint ENDPOINT_MDW = new ApiEndpoint(true, "mdw", 31445, false);
        public static ApiEndpoint ENDPOINT_MDR = new ApiEndpoint(true, "mdr", 31445, false);
        public static ApiEndpoint ENDPOINT_MDIDX = new ApiEndpoint(true, "mdidx", 31446, false);        

        public string WebSocketAddress { get; set; }

        public string InstanceId { get; set; }

        public ApiClient(string subscriptionKey)
        {            
            _environment = ApiEnvironment.DEVELOPMENT;

            _subscriptionKey = subscriptionKey;

            WebSocketAddress = null;
            InstanceId = null;
        }

        public ApiClient(string subscriptionKey, string bearerToken)
            : this(subscriptionKey)
        {
            _bearerToken = bearerToken;
        }

        public ApiClient(string subscriptionKey, ApiEnvironment environment)
        {
            _subscriptionKey = subscriptionKey;
            _environment = environment;

            WebSocketAddress = null;
            InstanceId = null;
        }

        public ApiClient(string subscriptionKey, ApiEnvironment environment, string bearerToken)
            : this(subscriptionKey, environment)
        {            
            _bearerToken = bearerToken;            
        }

        public ApiRequest CreateApiRequest(string brandUuid)
        {
            return new ApiRequest(this, brandUuid);
        }

        internal string GetBearerToken()
        {
            return _bearerToken;
        }

        internal async Task<byte[]> ProcessAsync(bool apiManagement, ApiEndpoint apiEndpoint, string brandUuid, 
            string transactionUuid, string url, byte[] gpbContent, string[] binaries, Method method, CancellationToken? cancellationToken = null)
        {
            string apiUrl = apiEndpoint.GetApiEndpointUrl(_environment, apiManagement);

            apiUrl += url;

            _httpClient = new HttpClient();
            _httpClient.Timeout = TimeSpan.FromMinutes(2);

            _httpClient.DefaultRequestHeaders.Accept.Clear();
            _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/x-protobuf"));            

            if (!string.IsNullOrEmpty(_bearerToken))
                _httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _bearerToken);

            if (!string.IsNullOrEmpty(brandUuid))
                _httpClient.DefaultRequestHeaders.Add(ApiConstants.REQUEST_HEADER_BRAND_UUID, brandUuid);

            if (!string.IsNullOrEmpty(WebSocketAddress))
                _httpClient.DefaultRequestHeaders.Add(ApiConstants.REQUEST_HEADER_WEBSOCKET_ADDRESS, WebSocketAddress);

            if (!string.IsNullOrEmpty(transactionUuid))
                _httpClient.DefaultRequestHeaders.Add(ApiConstants.REQUEST_HEADER_TRANSACTION_UUID, transactionUuid);

            if (!string.IsNullOrEmpty(_subscriptionKey))
                _httpClient.DefaultRequestHeaders.Add(ApiConstants.REQUEST_HEADER_SUBSCRIPTION_KEY, _subscriptionKey);

            if (!string.IsNullOrEmpty(InstanceId))
                _httpClient.DefaultRequestHeaders.Add(ApiConstants.REQUEST_HEADER_INSTANCE_ID, InstanceId);

            _httpClient.DefaultRequestHeaders.Add(ApiConstants.REQUEST_HEADER_CONTIDIO_SDK, ApiConstants.CONTIDIO_SDK_IDENTIFIER);

            HttpResponseMessage response = null;

            try
            {                
                if (binaries == null || binaries.Length == 0)
                {
                    HttpContent httpContent = null;

                    try
                    {
                        if (gpbContent != null)
                        {
                            httpContent = new ByteArrayContent(gpbContent);
                        }
                        else
                        {
                            httpContent = new ByteArrayContent(new byte[] { });
                        }

                        switch (method)
                        {
                            case Method.GET:
                                if (cancellationToken != null)
                                    response = await _httpClient.GetAsync(apiUrl, (CancellationToken)cancellationToken)
                                        .ConfigureAwait(false);
                                else
                                    response = await _httpClient.GetAsync(apiUrl)
                                        .ConfigureAwait(false);
                                break;
                            case Method.POST:
                                if (cancellationToken != null)
                                    response = await _httpClient.PostAsync(apiUrl, httpContent, (CancellationToken)cancellationToken)
                                        .ConfigureAwait(false);
                                else
                                    response = await _httpClient.PostAsync(apiUrl, httpContent)
                                        .ConfigureAwait(false);
                                break;
                            case Method.PUT:
                                if (cancellationToken != null)
                                    response = await _httpClient.PutAsync(apiUrl, httpContent, (CancellationToken)cancellationToken)
                                        .ConfigureAwait(false);
                                else
                                    response = await _httpClient.PutAsync(apiUrl, httpContent)
                                        .ConfigureAwait(false);
                                break;
                            case Method.DELETE:
                                HttpRequestMessage request = new HttpRequestMessage
                                {
                                    Content = httpContent,
                                    Method = HttpMethod.Delete,
                                    RequestUri = new Uri(apiUrl)
                                };

                                if (cancellationToken != null)
                                    response = await _httpClient.SendAsync(request, (CancellationToken)cancellationToken)
                                        .ConfigureAwait(false);
                                else
                                    response = await _httpClient.SendAsync(request)
                                        .ConfigureAwait(false);
                                break;
                            default:
                                throw new InternalErrorBackendException("Invalid HTTP method found");
                        }
                    }
                    catch (Exception e)
                    {
                        throw e;
                    }
                    finally
                    {
                        try
                        {
                            if (httpContent != null)
                                httpContent.Dispose();
                        }
                        catch (Exception)
                        {
                            // ignore...
                        }
                    }
                } else
                {
                    if (method != Method.POST && method != Method.PUT)
                        throw new InternalErrorBackendException("Invalid HTTP method found");

                    MultipartFormDataContent multipartContent = null;

                    try
                    {
                        multipartContent = new MultipartFormDataContent("Upload----" + DateTime.Now.ToString(CultureInfo.InvariantCulture));

                        string gpbContentEncoded = null;

                        if (gpbContent != null)
                            gpbContentEncoded = Convert.ToBase64String(gpbContent);
                        else
                            gpbContentEncoded = Convert.ToBase64String(new byte[] { });

                        multipartContent.Add(new StringContent(gpbContentEncoded), "gpb");                        

                        foreach (string binary in binaries)
                        {
                            FileStream fileStream = null;

                            try
                            {
                                if (!File.Exists(binary))
                                    throw new InternalErrorBackendException("Binary with path '" + binary + " was not found");                                

                                fileStream = File.OpenRead(binary);
                                var streamContent = new StreamContent(fileStream);
                                streamContent.Headers.Add("Content-Type", "application/octet-stream");
                                streamContent.Headers.Add("Content-Disposition", "form-data; name=\"files\"; filename=\"" +
                                    Path.GetFileName(binary) + "\"");

                                multipartContent.Add(streamContent, "files", Path.GetFileName(binary));
                            } catch (Exception)
                            {
                                try { 
                                    if (fileStream != null)
                                        fileStream.Dispose();
                                }
                                catch (Exception)
                                {
                                    // ignore...
                                }
                            }
                        }

                        switch (method)
                        {
                            case Method.POST:
                                if (cancellationToken != null)
                                    response = await _httpClient.PostAsync(apiUrl, multipartContent, (CancellationToken)cancellationToken)
                                        .ConfigureAwait(false);
                                else
                                    response = await _httpClient.PostAsync(apiUrl, multipartContent)
                                        .ConfigureAwait(false);
                                break;
                            case Method.PUT:
                                if (cancellationToken != null)
                                    response = await _httpClient.PutAsync(apiUrl, multipartContent, (CancellationToken)cancellationToken)
                                        .ConfigureAwait(false);
                                else
                                    response = await _httpClient.PutAsync(apiUrl, multipartContent)
                                        .ConfigureAwait(false);
                                break;
                            default:
                                throw new InternalErrorBackendException("Invalid HTTP method found");
                        }
                    }
                    catch (Exception e)
                    {
                        throw e;
                    }
                    finally
                    {
                        try
                        {
                            if (multipartContent != null)
                                multipartContent.Dispose();
                        }
                        catch (Exception)
                        {
                            // ignore...
                        }
                    }
                }
                
                if (response.StatusCode == HttpStatusCode.NoContent)
                {
                    _handleAsError = false;
                    return new byte[] { };
                }

                if (!response.IsSuccessStatusCode)
                {
                    _handleAsError = true;
                    _errorStatusCode = response.StatusCode;
                    _errorReason = response.ReasonPhrase;
                }
                else
                {
                    _handleAsError = false;
                }

                byte[] result = await response.Content.ReadAsByteArrayAsync()
                    .ConfigureAwait(false);

                if (_handleAsError)
                {
                    if (GetResponseIsProtobuf(response))
                    {
                        Error error = Error.ParseFrom(result);

                        throw BackendException.GetTransientExceptionFor(error);
                    } else
                    {
                        throw new InternalErrorBackendException(
                            "The server responded with HTTP error status code " + _errorStatusCode + " (" + 
                                (string.IsNullOrEmpty(_errorReason) ? "unknown reason" : _errorReason) + ")");
                    }
                }
                else
                {
                    return result;
                }
            }
            catch (Exception e)
            {
                throw e;
            }
            finally
            {
                try
                {
                    if (response != null)
                        response.Dispose();
                }
                catch (Exception)
                {
                    // ignore...
                }
            }
        }

        private bool GetResponseIsProtobuf(HttpResponseMessage response)
        {
            if (response.Content.Headers != null && response.Content.Headers.ContentType != null)
            {
                string contentType = response.Content.Headers.ContentType.MediaType;
                return contentType.StartsWith("application/x-protobuf");
            }

            return false;
        }        

        public void Dispose()
        {
            if (_httpClient != null)
            {
                _httpClient.Dispose();
                _httpClient = null;
            }
        }
    }
}

File: Java/src/main/java/com/contidio/sdk/api/ApiClient.java
package com.contidio.sdk.api;

import java.io.File;
import java.util.Base64;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ExecutionException;

import org.apache.commons.httpclient.HttpStatus;
import org.apache.commons.lang3.StringUtils;
import org.asynchttpclient.AsyncHttpClient;
import org.asynchttpclient.BoundRequestBuilder;
import org.asynchttpclient.Response;
import org.asynchttpclient.request.body.multipart.FilePart;
import org.asynchttpclient.request.body.multipart.StringPart;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.contidio.sdk.api.eventbus.ApiEventBus;
import com.contidio.sdk.api.eventbus.ApiEventBusCallback;
import com.contidio.sdk.api.eventbus.ApiEventBusEventProcessor;
import com.contidio.sdk.exceptions.BackendException;
import com.contidio.sdk.exceptions.InternalErrorBackendException;

public class ApiClient {
	private static final Logger logger = LoggerFactory.getLogger(ApiClient.class);
	
	public enum Method
    {
        GET,
        PUT,
        POST,
        DELETE
    }
	
	private boolean handleAsError;
	private String bearerToken;
	
	private int errorStatusCode;
	private String errorReason;
	
	private ApiEnvironment environment;
	
	private String subscriptionKey;
	
	public static ApiEndpoint ENDPOINT_WWW = new ApiEndpoint(true, "www", 31442, true);
	public static ApiEndpoint ENDPOINT_ID = new ApiEndpoint(true, "id", 31443, false);
	public static ApiEndpoint ENDPOINT_CW = new ApiEndpoint(true, "cw", 31444, false);
	public static ApiEndpoint ENDPOINT_CR = new ApiEndpoint(true, "cr", 31444, false);
	public static ApiEndpoint ENDPOINT_MDW = new ApiEndpoint(true, "mdw", 31445, false);
	public static ApiEndpoint ENDPOINT_MDR = new ApiEndpoint(true, "mdr", 31445, false);
	public static ApiEndpoint ENDPOINT_MDIDX = new ApiEndpoint(true, "mdidx", 31446, false);
	
	private String webSocketAddress;
	
	private AsyncHttpClient asyncHttpClient;
	
	private ApiEventBus apiEventBus;
	
	private Object apiEventBusMutex = new Object();
	
	private String instanceId;
	
	public ApiClient(AsyncHttpClient asyncHttpClient, String subscriptionKey) {
		environment = ApiEnvironment.DEVELOPMENT;
		
		this.subscriptionKey = subscriptionKey;
		
		this.webSocketAddress = null;
		this.instanceId = null;
		
		this.asyncHttpClient = asyncHttpClient;
	}
	
	public ApiClient(AsyncHttpClient asyncHttpClient, String subscriptionKey, String bearerToken) {
		this(asyncHttpClient, subscriptionKey);
				
		this.bearerToken = bearerToken;
	}
	
	public ApiClient(AsyncHttpClient asyncHttpClient, String subscriptionKey, ApiEnvironment environment) {
		this.subscriptionKey = subscriptionKey;
		this.environment = environment;
		
		this.webSocketAddress = null;
		
		this.asyncHttpClient = asyncHttpClient;
	}
	
	public ApiClient(AsyncHttpClient asyncHttpClient, String subscriptionKey, ApiEnvironment environment, String bearerToken) {
		this(asyncHttpClient, subscriptionKey, environment);
		
		this.bearerToken = bearerToken;
	}
	
	public String getWebSocketAddress() {
		return webSocketAddress;
	}

	public void setWebSocketAddress(String webSocketAddress) {
		this.webSocketAddress = webSocketAddress;
	}

	public ApiRequest createRequest(String brandUuid) {
		return new ApiRequest(this, brandUuid);
	}
	
	public String getBearerToken() {
		return bearerToken;
	}
	
	public void setInstanceId(String instanceId) {
		this.instanceId = instanceId;
	}
	
	public String getInstanceId() {
		return instanceId;
	}
	
	byte[] process(boolean apiManagement, ApiEndpoint apiEndpoint, String brandUuid, String transactionUuid,
			String url, byte[] gpbContent, File[] binaries, Method method)
		throws Exception
	{
		String apiUrl = apiEndpoint.getApiEndpointUrl(environment, apiManagement);
		
		apiUrl += url;
				
		BoundRequestBuilder requestBuilder = null;
		
		if (binaries == null || binaries.length == 0) {
			switch(method) {
				case GET:
					requestBuilder = asyncHttpClient.prepareGet(apiUrl);
					break;
				case POST:
					requestBuilder = asyncHttpClient.preparePost(apiUrl);
					
					if (gpbContent != null)
						requestBuilder.setBody(gpbContent);
					else
						requestBuilder.setBody(new byte[] {});
					
					break;
				case PUT:
					requestBuilder = asyncHttpClient.preparePut(apiUrl);
					
					if (gpbContent != null)
						requestBuilder.setBody(gpbContent);
					else
						requestBuilder.setBody(new byte[] {});
					
					break;
				case DELETE:
					requestBuilder = asyncHttpClient.prepareDelete(apiUrl);
					
					if (gpbContent != null)
						requestBuilder.setBody(gpbContent);
					else
						requestBuilder.setBody(new byte[] {});
					
					break;
				default:
					throw new InternalErrorBackendException("Invalid HTTP method found");
			}
		} else {
			String gpbContentEncoded = null;
			if (gpbContent != null)
				gpbContentEncoded = Base64.getEncoder().encodeToString(gpbContent);
			else
				gpbContentEncoded = Base64.getEncoder().encodeToString(new byte[] {});
			
			switch(method) {
				case POST:
					requestBuilder = asyncHttpClient.preparePost(apiUrl);
										
					requestBuilder.addBodyPart(new StringPart("gpb", gpbContentEncoded));
										
					break;
				case PUT:
					requestBuilder = asyncHttpClient.preparePut(apiUrl);
					
					requestBuilder.addBodyPart(new StringPart("gpb", gpbContentEncoded));					
					
					break;
				default:
					throw new InternalErrorBackendException("Invalid HTTP method found");
			}
			
			for (File binary : binaries) {
				if (!binary.exists())
					throw new InternalErrorBackendException("Binary with path '" + binary.getAbsolutePath() + " was not found");
				
				if (!binary.canRead())
					throw new InternalErrorBackendException("Binary with path '" + binary.getAbsolutePath() + " cannot be read");
				
				requestBuilder.addBodyPart(new FilePart("files", binary));
			}
		}
		
		requestBuilder.addHeader("Accept", "application/x-protobuf");
		
		if (!StringUtils.isEmpty(bearerToken))
			requestBuilder.addHeader("Authorization", "Bearer " + bearerToken);	
		
		if (!StringUtils.isEmpty(brandUuid))
			requestBuilder.addHeader(ApiConstants.REQUEST_HEADER_BRAND_UUID, brandUuid);
		
		if (!StringUtils.isEmpty(webSocketAddress))
			requestBuilder.addHeader(ApiConstants.REQUEST_HEADER_WEBSOCKET_ADDRESS, webSocketAddress);
		
		if (!StringUtils.isEmpty(transactionUuid))
			requestBuilder.addHeader(ApiConstants.REQUEST_HEADER_TRANSACTION_UUID, transactionUuid);
		
		if (!StringUtils.isEmpty(subscriptionKey))
			requestBuilder.addHeader(ApiConstants.REQUEST_HEADER_SUBSCRIPTION_KEY, subscriptionKey);
		
		if (!StringUtils.isEmpty(instanceId))
			requestBuilder.addHeader(ApiConstants.REQUEST_HEADER_INSTANCE_ID, instanceId);
		
		requestBuilder.addHeader(ApiConstants.REQUEST_HEADER_CONTIDIO_SDK, ApiConstants.CONTIDIO_SDK_IDENTIFIER);
		
		Response response = null;
		
		try {
			response = requestBuilder.execute().get();
			
			int statusCode = response.getStatusCode();
			
			if (statusCode == HttpStatus.SC_NO_CONTENT) {
				handleAsError = false;
				return new byte[] { };
			}
			
			if (statusCode != HttpStatus.SC_OK) {
				handleAsError = true;
				errorStatusCode = response.getStatusCode();
				errorReason = response.getStatusText();
			} else {
				handleAsError = false;
			}
			
			byte[] result = response.getResponseBodyAsBytes();
			
			if (handleAsError) {
				if (getResponseIsProtobuf(response)) {
					com.contidio.sdk.proto.MetadataProtocol.Error error = com.contidio.sdk.proto.MetadataProtocol.Error
							.parseFrom(result);
					
					throw BackendException.getTransientExceptionFor(error);						
				} else {
					throw new InternalErrorBackendException(
							"The server responded with HTTP error status code " + errorStatusCode + " (" + 
								(StringUtils.isEmpty(errorReason) ? "unknown reason" : errorReason) + ")");
				}
			} else {
				return result;
			}
		} catch (Exception e) {
			throw e;
		} finally {			
			// nothing to do
		}		
	}

	private boolean getResponseIsProtobuf(Response response) {
		if (response.getContentType() != null) {
			String contentType = response.getContentType();
			return StringUtils.startsWith(contentType, "application/x-protobuf");
		}
		
		return false;
	}
	
	public void connectToEventBus(ApiEventBusCallback apiEventBusCallback) 
		throws InterruptedException, ExecutionException 
	{
		connectToEventBus(apiEventBusCallback, null, null);
	}
	
	public void connectToEventBus(
			ApiEventBusCallback apiEventBusCallback, 
			ApiEventBusEventProcessor apiEventBusEventProcessor) 
		throws InterruptedException, ExecutionException
	{
		connectToEventBus(apiEventBusCallback, apiEventBusEventProcessor, null);
	}
		
	public void connectToEventBus(
			ApiEventBusCallback apiEventBusCallback, 
			ApiEventBusEventProcessor apiEventBusEventProcessor,
			Set<String> webSocketAddresses) 
		throws InterruptedException, ExecutionException 
	{
		logger.info("Connecting to event bus");
		
		synchronized (apiEventBusMutex) {
			if (StringUtils.isEmpty(webSocketAddress))
				webSocketAddress = UUID.randomUUID().toString();
			
			if (webSocketAddresses == null)
				webSocketAddresses = new HashSet<String>();
			
			webSocketAddresses.add(webSocketAddress);
			
			disconnectFromEventBus();
			
			String endpoint = ENDPOINT_MDIDX.getApiEndpointUrl(environment, false);
									
			apiEventBus = new ApiEventBus(
					asyncHttpClient, 
					endpoint, 
					apiEventBusCallback, 
					apiEventBusEventProcessor,
					webSocketAddresses);
			
			apiEventBus.connect();
		}
		
		logger.info("Connected to event bus");
	}
	
	public void disconnectFromEventBus() {
		logger.info("Disconnecting from event bus");
		
		synchronized (apiEventBusMutex) {
			if (apiEventBus != null) {
				apiEventBus.disconnect();
				
				apiEventBus = null;
			}
		}
		
		logger.info("Disconnected from event bus");
	}
}
